### PR TITLE
Text: Implement Text module

### PR DESCRIPTION
**Pull request by YakoYakoYokuYoku**
# Text module

This is a proposal for a text module in FreeCAD. Maybe this can hit the feature window for `1.1` if it gets through in time.

## Rationale

Since the invention of the moveable-type printing press the world of typography went to a whole new level, paving up the path towards what we know today. Since then many metrics and standards were developed for printing and displaying text in many parts of our everyday life. With the introduction of computers things took the next step. Typography is undeniably an important technique of the contemporany world.

And there lies the reason of this PR, typography is an art by itself, just like solid modelling or technical drawing. Prior to this, if you want to add text to your parts in FreeCAD you have to use the Draft workbench, which is kinda limited in this respect. Sure, you can emboss or engrave text, but if it contains characters from different writing systems, it has to be put on a curve or a shape, it has to be warped or, as crazy as it sounds, if it has to use a certain font but you only have its name then you are pretty much on your own or you have to rely on a not so trivial solution. This workbench aims to give a proper response to those kind of situations.

Some potential users could be #22217 and the tomate44/CurvesWB addon.

Text prior to the workbench to module only revision.

<details>
One could ask why not doing it in Sketcher or adding more features in Draft, then my answer is that there are many limitations to such alternatives. Namely, although typeset text is constrained (which is what Sketcher does with geometry), much of that is decided by the font designer and not by the user, which makes sense. Also, working in Sketcher with a string of text is pretty involved and error prone. On the other hand, if we were to add more to Draft we are effectively creeping it. So that's why in my opinion that relegating typesetting into its own workbench is the best option out there.
</details>

## Features

The Text module will, ideally, do many things.

- Shaping strings in a FreeCAD part with a new `ShapeText` object.
- Managing fonts for said strings, like if they were sourced from a family name, a file or a document object.
- Placing text on curves, planes and surfaces.
- Warping text.
- Handling complex text layouts, like mixing different writing systems and fonts.
- Etc.

## Testing

This was tested by engraving and embossing text on different parts.

## Images

image
image

## Videos

[Screencast_20240930_185103.webm](https://github.com/user-attachments/assets/8bb19286-5be3-4b94-9fea-347d67dd5baf)

### CODE REVIEW

**Comment by benj5378** [comment]
Early Christmas present! Just curious, how much time did you spend working on this?

**Comment by macdroid53** [comment]
Can it handle text with beveled top edges?
Can it handle draft of the text (as in the draft of a mold to allow extraction after casting)?

**Comment by pierreporte** [comment]
I don’t know if a dedicated workbench is the best option. Generic commands maybe but the buttons could be placed in the toolbar of various workbenches instead.

@FreeCAD/design-working-group @FreeCAD/cad-advisory-group what do you think?

**Comment by macdroid53** [comment]
I not sure if this addresses the issues that are typically associated with doing text in FreeCAD. 
Given a valid font, the positioning of the Shapestring when working in Part workbench or in Part Design are typically the major issues people have.
Pocket through causing multiple solids has been addressed in Part Design with the relaxing of the single solid rule.
The other major issue people have is draft for mold release and bevel/fillet.

**Comment by YakoYakoYokuYoku** [comment]
> Do you have screencasts or clips of you using it? Also have you considered making this into a 3rd party external workbench ?

Yes, I've added it to the PR comment. And yes, I've considered the idea, the issue is that because it uses OpenCascade classes that aren't interfaced by Part I've discarded it. I could've also used pythonocc-core or another Python binding for OCCT, but I see it as too cumbersome.

<details>
[Screencast_20240930_185103.webm](https://github.com/user-attachments/assets/8bb19286-5be3-4b94-9fea-347d67dd5baf)
</details>

> Early Christmas present! Just curious, how much time did you spend working on this?

Since the last month, but I've thought of the idea way back in time.

> Can it handle text with beveled top edges? Can it handle draft of the text (as in the draft of a mold to allow extraction after casting)?

I think that I might need an explanation on what's this, as of now I don't know what you might be referring to.

> I don’t know if a dedicated workbench is the best option. Generic commands maybe but the buttons could be placed in the toolbar of various workbenches instead.

And those workbenches could use the functionality from the Text module, just saying.

**Comment by macdroid53** [comment]
Basically fillets and chamfers of top or bottom edges.

**Comment by yorikvanhavre** [comment]
That's impressive work :)

My observations so far (haven't compiled or tried):

- I understand your rationale about not using Draft or Sketches, it makes sense
- The code is properly organized into one separate module
- Yet, it basically proposes one single tool, similar to the Draft ShapeString tool. In fact, it could maybe completely replace Draft ShapeString.
- This could actually become a built-in text tool for FreeCAD. It could not provide any workbench, but provides a Text tool, to be added to, why not, both Draft and PartDesign

One detail I specially liked, is that you are able to pick a font by name OR by file. Would it be possible to use the standard Qt QFontChooser widget for choosing the font? It is nicer and displays a preview of the font. It is however tricky to match it to a ttf file, see https://github.com/FreeCAD/FreeCAD/issues/9458 , but you seem to have solved that problem already...

**Comment by luzpaz** [comment]
Note: tickets that will be effected from this and effecting this feature would be:  
* #5746
* #12944  
* #12844

**Comment by qewer33** [comment]
> This could actually become a built-in text tool for FreeCAD. It could not provide any workbench, but provides a Text tool, to be added to, why not, both Draft and PartDesign

I agree with this point on yorik's comment. One could argue FreeCAD already has "too many" built-in workbenches for specific stuff, adding another one just for Text doesn't make sense IMHO.

On the other hand, seeing an advanced text tool like this one as a "global command" would be amazing. This way we can add it to the toolbars of workbenches that need it like Part, Part Design and Draft. And in the future it can have more advanced stuff in the same task dialog such as different "placement modes" (eg. Flat, Follow Path, Map to Surface etc.) all neatly organized.

Also on an unrelated note: the task dialog definitely needs a preview option.

**Comment by adrianinsaval** [comment]
+1 on the point that a whole workbench for a text tool is overkill

**Comment by hyarion** [comment]
Adding a new, built in, workbench for text doesn't feel right to me either, we have a lot of them already.

Additionally, if we where to add this new workspace, it would force us to be backwards compatible with it for quite some time. I would strongly suggest that this was made into an addon instead. If occt functions needs to be exported then that is less of a problem. (Adding features takes only a small fraction of the time it takes to maintain it during its life)


I still think text should be part of sketcher so text can be constrained positionally, rotationally and in scale (as a whole, not glyph by glyph).

It would be good to see how other cad software have handled text, maybe we can draw inspiration from there?

**Comment by obelisk79** [comment]
The functionality appears to be excellent. But I'll have to chime in and agree that this shouldn't be a workbench. A global function that can be called/added to workbench toolbars as applicable seems more logical to me.

**Comment by hyarion** [comment]
I found this old forum thread regarding text in PD vs Sketcher. It's a good thread to read that is closely related to this PR. There, at the end, abdullah and davidosterberg discuss how to implement it in Sketcher instead of a separate PD tool.

**Comment by kadet1090** [comment]
This is similar to a concept that I commented here: https://github.com/FreeCAD/FreeCAD/issues/15151#issuecomment-2395506134 - then we could implement it in sketcher without being to obnoxious. We cannot really do letters as real geometry in sketch as this would be a nightmare to solver and to move around. Basically it for sketch we should only be able to constraint arbitrary "anchor" / "origin" point that we can choose from various options (like visual center, baseline center, baseline left etc. This would make solving the sketch trivial as we basically deal with one point only. Not to say that implementing it that way could bring us a lot more useful features in sketcher as we now lock ability to have such "exotic" geometry.

However, this is not something that should be part of this PR a tremendous task to do. For now I think that proposed "Text" object seems rather fine - but as other noted, it should be a global tool available from various workbenches. 

Some tools are also quite odd: 
1. why do we need separate "Edit" tool? Double clicking on existing text should be enough.
2. from what I see "Map Text To Face" simply changes attachment with in a different way - launching the attachment dialog should be enough
3. "Map Text to Face" does not work with curved surfaces - at least I was not able to make it work
4. I cannot figure out how to position text relative to a face - it seems to be always attached to origin, which is not something that I want in most use cases

To sum up - text tool seems solid, but the rest should be done more consistently with what FreeCAD offers and not reinvent wheel.

**Comment by PaddleStroke** [comment]
- +1 against new workbench. You can keep a text module if necessary (though I'd say it probably belong in Part), but then the tool should be added where it belongs: Draft, PartDesign...
- Text editing task dialog should belong to core. I say this because several workbench needs text editing in different context : This and TechDraw (annotation tool) at least. And even though the implementation of the graphics have nothing in common, the handling of text input and formating is identical. So basically we would need a TaskText.cpp in core that would be a HTML formatter basically. Then the separate tools in Techdraw and Part reimplement this to handle HTML text -> something useful in the context.
- The sketcher idea seems nice. But as Kadet mentioned this is asking a lot from this PR

**Comment by FlachyJoe** [comment]
What about basing the Text object on SVG?
The new workbench could be a 2d path editor including text but also basic SVG features with ability to open the path in an another app e.g. inkscape.

**Comment by FlachyJoe** [comment]
@macdroid53 **In order** it works well :smiley:

**Comment by maxwxyz** [comment]
Any news on the PR? CI is failing too.

**Comment by maxwxyz** [comment]
@YakoYakoYokuYoku ping

**Comment by YakoYakoYokuYoku** [comment]
> Any news on the PR? CI is failing too.

Was occupied with some stuff irl, probably will take a look in the following week. If I manage to I'll add font document objects, that is, a way to embed fonts in a FreeCAD file. Pretty cool, huh?

On the other hand is that I have no clue on how to preview text when editing like what Sketcher does, if someone could give a clue that'd be very appreciated.

**Comment by macdroid53** [comment]
> @macdroid53 **In order** it works well 😃
From the number of people I see on the forum, Reddit, Facebook, and Discord asking "why does this imported SVG not <insert any number of things here>" I would not say SVG works _well_ in FreeCAD.

**Comment by maxwxyz** [comment]
what is the verdict on this PR? There are also some conflicts.

**Comment by hyarion** [comment]
> what is the verdict on this PR? There are also some conflicts.

From what I read in the comments above, the general feedback is that another workbench with only one tool isn't such a good idea.

---

Regarding a path to get text into freecad:
Feature wise, adding a text tool is great, but personally, I don't see that it is a good idea to add it as a one tool workbench.
I think it should be implemented either:
* in the sketcher, or
* as an addon

Doing either of the above would be a great improvement of freecad.

Regarding "asking too much of the PR", while it is true that it would have been a lot of work, the work could have been reduced by having the discussion before the feature was implemented.

If we need a verdict, maybe we could add it onto the next dev meeting?

**Comment by yorikvanhavre** [comment]
I think it is still very worth going forward with this... The code and functionality are good. Maybe what we need is just decide on the UI side of things, if this needs a separate workbench or be included in other workbenches. But don't feel discouraged because of our internal UI debates @YakoYakoYokuYoku ;)

**Comment by oursland** [comment]
I would love to see this added to the Sketcher, personally.

**Comment by Roy-043** [comment]
Another development: #22217.

**Comment by YakoYakoYokuYoku** [comment]
Thanks for the heads up.

**Comment by YakoYakoYokuYoku** [comment]
This has now been reduced into a simple Python module, if it's desirable to expand it back into a proper workbench then it could be done in another PR. Some features could be added by toying around with what OpenCascade provides.

**Comment by hyarion** [comment]
To be honest, I don't know what to do with this one. A workbench was a bit much for a single tool, but a module is a bit overkill too.
The functionality is nice, and it is quite cleanly built, but it would be good if this could be part of `Part` instead, in a way that both Draft and Sketcher could make use of it too.
I don't have any good answers for how to proceed, but trying to implement text in sketcher in a clean way that also refactors draft would probably bring some insights on how to proceed here.

**Comment by pierreporte** [comment]
Is it really overkill though? Maybe this module could be used by other parts of the programs later, not just those that depend from Part. Do Draft and Sketcher depend on Part?

**Comment by hyarion** [comment]
> Is it really overkill though? Maybe this module could be used by other parts of the programs later, not just those that depend from Part. Do Draft and Sketcher depend on Part?

Yes they do. Most parts that interact with shapes depends on Part.

**Comment by MisterMakerNL** [comment]
Just implements the method that works, you can always make it better.

**Comment by chennes** [comment]
We discussed this in the PR Review Meeting this morning and the general consensus is that as it stands now, it really doesn't make sense as a separate module. So we'd like to figure out what the long-term plan is here. How much more code do you see this module gaining over time?

**Comment by adrianinsaval** [comment]
it was suggested that this should be included in Part instead if it's not expected to grow a lot

**Comment by YakoYakoYokuYoku** [comment]
> So we'd like to figure out what the long-term plan is here. How much more code do you see this module gaining over time?

It depends, if we go beyond these basic features from this PR, then a module would be required to keep things organized. People may want to do text along a path. People may want to transform text. People may want to mix fonts, which is not uncommon in these times with emojis or multiple languages in a string. People may have to rely on a text engine that handles text direction properly.

All of that requires an API that might not fit in Part by itself due to the fact that it carries a certain kind of complexity alien to that workbench and that their methodologies differ between them. Part, in summary, manages geometries and operations on them but Text should handle strings with a capability on a per character manipulation plus dealing with any font involved in the typesetting of text.

While it might not materialize itself as a workbench, Text may indeed grow in code size.

**Comment by Reqrefusion** [comment]
> > So we'd like to figure out what the long-term plan is here. How much more code do you see this module gaining over time?
> 
> It depends, if we go beyond these basic features from this PR, then a module would be required to keep things organized. People may want to do text along a path. People may want to transform text. People may want to mix fonts, which is not uncommon in these times with emojis or multiple languages in a string. People may have to rely on a text engine that handles text direction properly.
> 
> All of that requires an API that might not fit in Part by itself due to the fact that it carries a certain kind of complexity alien to that workbench and that their methodologies differ between them. Part, in summary, manages geometries and operations on them but Text should handle strings with a capability on a per character manipulation plus dealing with any font involved in the typesetting of text.
> 
> While it might not materialize itself as a workbench, Text may indeed grow in code size.

Yes, you're right; text can contain many elements. When you think about it the way you described, it makes sense as a module.

**Comment by PaddleStroke** [comment]
I have mixed feelings about this. Because even if there could be more code, there currently isn't. So we may end up in the same situation as with the 'Inspection' or the late 'Image' workbenches, that were created with the same intent, but that both ended up never getting anymore code even after 12 years.

So it is kind of putting the cart before the horse. Moving the code to a separate addon should happen when the quantity of code becomes so big that it makes sense.

Besides, regarding text along a path and text transform, I think both of these operation are probably standard transform operation. So that would rely on transform code rather than on text code specifically no? And 'handling text direction' I am not sure what this means?

When you talk about mixed font, what I think about is HTML. As current tools only take raw strings to path. It would indeed be a great feature to be able to take HTML formated text to path.

Long story short, I think the question is, do you plan on tackling those features in the near future? If not then that PR is a lot of boiler plate code and project complexity for what could be a single class/file in Part module.

**Comment by hyarion** [comment]
I agree with paddle here, moving this to a separate module seems premature. I think we should be able to integrate this in Part. If it grows too much then we can move it out to a new module when needed.

**Comment by maxwxyz** [comment]
@YakoYakoYokuYoku do you plan on migrating this to Part?

**Comment by pierreporte** [comment]
Maybe it would be better in Part, but the PR already exists in future-proof state. Is it really worth it to change again?

**Comment by hyarion** [comment]
I think the problem here is that we're not 100% sure on what we want yet.

It would be better to have it in Part yes, but this type of PRs easily becomes a wild goose hunt. If OP moves it, will it be accepted then? I wouldn't count on it myself, there will most likely be more feedback then..

One problem here is that building a module which _might_ be good to have in the future adds a lot of boilerplate which right now can be avoided. The same thing is to have a text workbench, it _may_ be good in the future, but right now it isn't a good choice.

I would really like to see a path forward where sketcher gets text support, but where the code for generating the geometry is implemented like in this PR but located in Part. But right now that falls between two PRs and merging support for what _might_ be needed is difficult without the actual implementation that shows that it _is_ needed.

**Comment by YakoYakoYokuYoku** [comment]
> do you plan on migrating this to Part?

Only if it remains as an experimental feature for feedback so that version compatibility issues are avoided, otherwise a big no.

**Comment by chennes** [comment]
We discussed this in the review meeting this morning, and we are going to put it on hold until we see where @PaddleStroke takes the Sketcher text PR.

**Comment by PaddleStroke** [comment]
I will solve conflicts and rebase soon that PR @chennes

**Comment by pierreporte** [comment]
The other PR has been rebased. @YakoYakoYokuYoku Are you still willing to work on this?

**Comment by YakoYakoYokuYoku** [comment]
Sorry for the inactivity, I've been working on a personal project of quite importance, which could also come very handy for FreeCAD. I'll keep the PR up to date for all that matters.

**Comment by pierreporte** [comment]
Do you think it’s ready to be merged? Then you could mark it ready for review. @chennes wants to merge #22217 soon and it requires this one to be perfect.

**Comment by YakoYakoYokuYoku** [comment]
Font file embedding is a skeleton as of now, but text wires are generated as expected when using the Python API. I can mark this as ready and implement font objects later.

**Comment by maxwxyz** [comment]
And news on this PR?

**Comment by YakoYakoYokuYoku** [comment]
> And news on this PR?

This was up for reviews since, maybe my wording was confusing.

**Comment by pierreporte** [comment]
@chennes now that the sketcher text tool is there, is this PR still on hold?
